### ISSUE TITLE
DBAL plugin fine-grained interface

### ISSUE THREAD

**Issue by krolexey** [ory/hydra#1627]
**Is your feature request related to a problem? Please describe.**

We want to add another datasource using DBAL plugin. But Registry interface that need to be implemented for that contains too much unrelated information. It looks like a registry of all Hydra core components starting from config, logging, tracing and metrics managers, then REST handlers for both public and admin APIs and finishing with registries for clients, oauth, consent and jwk modules.

In order to customize only datasource we have to copy-paste the rest Registry Hydra source code. That makes further plugin support a nightmare.

**Describe the solution you'd like**

We'd like to see some smaller fine-grained interfaces to implement only datasource.

**Comment by aeneasr** [comment]
This topic has come up time and time again, the original plugin interface was written to support a certain use case we no longer support.

Writing your own database adapter is **a very bad idea and serious security issue** because:

1. Our internal interfaces change often, at least with every schema change. Writing your own adapter means in 99.9% of the cases that you will stay with the version your adapter works for. This is, for obvious reasons, terrible.
2. The SQL datastore supports advanced logic such as `ON CASCADE` statements that are necessary to implement certain security-related workflows such as properly removing a chain of tokens when one of them is being revoked or a client is deleted or consent is withdrawn. We obviously have some tests for that in place internally, but not all of them are exposed and they also change depending on internal schedules.

Most "we want to roll our own adapter" boil down to statements such as: "well we use (insert k/v, document datastore here) and we don't want to operate another system". If that's the reason you want to implement a new datastore - don't. It's riskier, more work, and more longterm maintenance to keep everything up to date than it is to just spin up a Google Cloud SQL or Amazon RDS instance, and way, way more secure.

In conclusion, we will not change this and will most likely deprecate plugins alltogether at some point.
